ztest: shell: Ztest framework enhancement.

gbarkadiusz opened the pull request:
This pull request introduces two significant improvements to the ZTEST framework:

**New Macro: ZTEST_P**
Introduces the new macro, `ZTEST_P`, enabling the creation of parameterized tests. These tests can receive parameters at runtime.
Includes new tests located in tests/ztest/ztest_params to validate and ensure the correctness of the `ZTEST_P` macro.
**Enhanced ztest_shell Subcommands**
`run-testcase` and `run-testsuite`: Both subcommands now support an additional `-r` (repeat-iter) parameter, allowing specific test cases or test suites to be executed multiple times.
Expanded `run-testcase` command: Adds a new `-p` parameter to the `run-testcase` subcommand, enabling the passing of parameters to tests at runtime.

gbarkadiusz committed:
ztest: expand ztest_shell command

Expand ztest_shell subcommands `run-testcase`
and `run-testsuite` to include additional parameter `-r`
repeat-iter. This allows to run a particular testcase or
testsuite several times.

Signed-off-by: Arkadiusz Cholewinski <arkadiuszx.cholewinski@intel.com>

gbarkadiusz committed:
tests: ztest: Add ztest param test.

Introduces the test for testing the new functionality
of ztest shell.

Passing -r (repeat iter) parameter. This allows to run
a particula testcase or testsuite several times.
Passing -p (any value) parameter. This allows to provide
a parameter to a test at runtime.

Signed-off-by: Arkadiusz Cholewinski <arkadiuszx.cholewinski@intel.com>

nashif commented on tests/ztest/ztest_param/pytest/test_parameters.py:
```diff
@@ -0,0 +1,33 @@
+    lines = shell.exec_command('ztest run-testcase ztest_params::test_string_param -p Zephyr')
+    assert 'Passed string:Zephyr' in lines, 'expected response not found'
+
+@pytest.mark.parametrize("test_input", ["Zephyr", "RTOS", "Intel"])
```
put something else other than Intel please :-)

gbarkadiusz replied:
Done :) , after all checks I'll mark it as ready for review.

LukaszMrugala commented on subsys/testsuite/ztest/src/ztest.c:
```diff
@@ -1321,15 +1326,57 @@ static int cmd_shuffle(const struct shell *sh, size_t argc, char **argv)
+	struct getopt_state *state;
+	int opt;
+	static struct option long_options[] = {{"repeat_iter", required_argument, 0, 'r'},
+		{0, 0, 0, 0}};
```
Could change the second element to `{NULL, 0, NULL, 0}` as they are pointers. `0` and `NULL` of course are the same, but some people prefer to always use `NULL`s for pointers.

gbarkadiusz replied:
Done.

LukaszMrugala commented on subsys/testsuite/ztest/src/ztest.c:
```diff
@@ -1321,15 +1326,57 @@ static int cmd_shuffle(const struct shell *sh, size_t argc, char **argv)
 {
+	struct getopt_state *state;
+	int opt;
+	static struct option long_options[] = {{"repeat_iter", required_argument, 0, 'r'},
```
`0` could be changed to `NULL` to keep consistency (`flag` is a pointer).

gbarkadiusz replied:
Done.

LukaszMrugala commented on subsys/testsuite/ztest/src/ztest.c:
```diff
@@ -1321,15 +1326,57 @@ static int cmd_shuffle(const struct shell *sh, size_t argc, char **argv)
+	const char *shell_command = argv[0];
 
-	ztest_set_test_args(argv[1]);
+	if (opt_num == 1) {
+		ztest_set_test_args(argv[3]);
+	} else {
+		ztest_set_test_args(argv[1]);
+	}
```
I'd add a short comment here.

gbarkadiusz replied:
Added the short explanation.

LukaszMrugala replied:
The comment now explains the 'what' and not the 'why' of the operation.

LukaszMrugala commented on subsys/testsuite/ztest/src/ztest.c:
```diff
@@ -1321,15 +1326,57 @@ static int cmd_shuffle(const struct shell *sh, size_t argc, char **argv)
+	int opt;
+	static struct option long_options[] = {{"repeat_iter", required_argument, 0, 'r'},
+		{0, 0, 0, 0}};
+	int opt_index = 0;
+	int val;
+	int opt_num = 0;
+	void *param = NULL;
+	int repeat_iter = 1;
```
You could use comma-separated declarations here, e.g.
```c
int val;
int opt_index = 0, opt_num = 0, repeat_iter = 1;
void *param = NULL;
```
Of course, what to group up and what to leave separate is a preference, so that's a judgement call.

LukaszMrugala commented on tests/ztest/ztest_param/pytest/test_parameters.py:
```diff
@@ -0,0 +1,33 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+import pytest
+from twister_harness import  Shell
```
Sneaky double-space

gbarkadiusz replied:
Cleaned up.

LukaszMrugala commented on tests/ztest/ztest_param/pytest/test_parameters.py:
```diff
@@ -0,0 +1,33 @@
+import pytest
+from twister_harness import  Shell
+
+def test_run_all_shuffled(shell: Shell):
+    shell.exec_command('ztest shuffle -c 1 -s 1')
+
+def test_run_testcase_once(shell: Shell):
+    shell.exec_command('ztest run-testcase ztest_params::test_dummy')
+
+def test_run_testcase_twice(shell: Shell):
+    shell.exec_command('ztest run-testcase ztest_params::test_dummy -r 2')
+
+def test_run_testsuite_twice(shell: Shell):
+    shell.exec_command('ztest run-testsuite ztest_params -r 2')
+
+def test_run_testsuite_once(shell: Shell):
+    shell.exec_command('ztest run-testsuite ztest_params')
```
Tests that do not test for results are suboptimal, but at least they should guard us against fatal errors.

LukaszMrugala commented on tests/ztest/ztest_param/pytest/test_parameters.py:
```diff
@@ -0,0 +1,33 @@
+    lines = shell.exec_command('ztest run-testcase ztest_params::test_int_param -p 44')
+    assert 'Passed int value:44' in lines, 'expected response not found'
+
+def test_run_testcase_with_string_parameter(shell: Shell):
+    lines = shell.exec_command('ztest run-testcase ztest_params::test_string_param -p Zephyr')
+    assert 'Passed string:Zephyr' in lines, 'expected response not found'
```
This one seems to be fully redundant, as the next test covers this case; I'd remove it.

gbarkadiusz replied:
It shows how to use `-p` with one parameter, but as recommended, I've removed it.

LukaszMrugala commented on subsys/testsuite/ztest/src/ztest.c:
```diff
@@ -1321,15 +1326,57 @@ static int cmd_shuffle(const struct shell *sh, size_t argc, char **argv)
+			repeat_iter = val;
+			opt_num++;
+			break;
+		case 'p':
+			param = state->optarg;
+			opt_num++;
+			break;
```
It seems that current solution allows for only one parameter. That parameter is CLI-provided, so it is in practice limited to simple types like `char*`, `int`, `float`. What if I need multiple parameters? I don't think I can easily pass in a struct-like structure and neither can I use multiple `-p` flags, correct?

Is there a method I have overlooked, or is this a possible enhancement?

gbarkadiusz replied:
So far, I've provided enhancement for passing one parameter into instance of the test.
If necessary, I'll expand this functionality for passing more complex types of parameters or multiple parameters.
In this approach we're based on the shell. So, the valid question is, how would you like to pass more complex parameters via shell?

The last test case explains how to pass data set into the test. But you're right it's technically one parameter for one instance of the test.

gbarkadiusz replied:
PS: I'd like to avoid passing the hardcoded parameters (the parameters which are the part of the source of a test) . But technically you are able to pass more complex parameters using the fixtures in ztest. Please take a look at https://docs.zephyrproject.org/latest/develop/test/ztest.html#test-fixtures

LukaszMrugala replied:
> So far, I've provided enhancement for passing one parameter into instance of the test. If necessary, I'll expand this functionality for passing more complex types of parameters or multiple parameters. In this approach we're based on the shell. So, the valid question is, how would you like to pass more complex parameters via shell?
> 
> The last test case explains how to pass data set into the test. But you're right it's technically one parameter for one instance of the test.

One practical method, in my opinion, would be something like
`<ztest call> -p param1=param1value -p param2=param2value`
that then gets transformed into a `data` struct with two fields, `param1` and `param2` that have the desired values. `zassert_equal(data->param1, "param1value")` should pass. List-like solution (think Twister's `--platform` option) seems incorrect here.

However, there are probably many other approaches. I personally think that expansion of this solution to multiple parametrisation would best be handled in a separate PR. That's why I gave an approve.

gbarkadiusz replied:
I think that's quite complex, because while initializing the sub-command via `SHELL_CMD_ARG` you are informing how many optional arguments this particular sub-command is expecting.

LukaszMrugala reviewed (dismissed):
Mostly aesthetical comments, with one seeming limitation noted.

gbarkadiusz commented on doc/develop/test/ztest.rst:
```diff
@@ -46,10 +46,13 @@ Below is an example of a test suite using a predicate:
 * :c:macro:`ZTEST` ``(suite_name, test_name)`` - Which can be used to add a test by ``test_name`` to a
   given suite by ``suite_name``.
+* :c:macro:`ZTEST_P` ``(suite_name, test_name)`` - Add a parameterized test to a given suite by specifying
+  the suite_name and test_name. You can then access the passed parameter within
```
Changed.

fabiobaltieri commented:
hey, this needs a rebase

gbarkadiusz commented:
> hey, this needs a rebase

Thank you, WIP.

gbarkadiusz committed:
ztest: expand ztest functionality.

Expanding the ZTEST shell to include a command that
accepts additional parameters, such as ztest run_testcase
<suite:test> -p any_value.

Adding the new ZTEST_P macro, to create the
parameterized test.

Signed-off-by: Arkadiusz Cholewinski <arkadiuszx.cholewinski@intel.com>

gbarkadiusz commented:
@nashif Please take a look, and merge if possible.

Terezvent commented on subsys/testsuite/ztest/include/zephyr/ztest_test.h:
Hello @gbarkadiusz 

You forgot to update the function call on line 589, so the ztest_run_test_suite function would also need to be updated
This is causing problems in the test: tests/kernel/timer/timer_behavior/kernel.timer.timer

gbarkadiusz replied:
Fixed by https://github.com/zephyrproject-rtos/zephyr/pull/77508

WalkingTalkingPotato reviewed (commented):
Needs additional work. If no improvement is planned, at least remove ZTEST_P from documentation as it is not generally usable (only in specific scenario when fixture corruption is acceptable and the testing is invoked through the shell interface).
